django_mongodb/compiler.py

```diff
@@ -0,0 +1,291 @@
+        the standard_ordering (a boolean, needed for MongoDB "$natural"
+        ordering).
+        """
+        opts = self.query.get_meta()
+        ordering = (
+            self.query.order_by or opts.ordering
+            if self.query.default_ordering
+            else self.query.order_by
+        )
+
+        if not ordering:
+            return self.query.standard_ordering
+
+        field_ordering = []
+        for order in ordering:
+            if LOOKUP_SEP in order:
+                raise DatabaseError(
+                    "Ordering can't span tables on non-relational backends " "(%s)." % order
+                )
+            if order == "?":
+                raise DatabaseError("Randomized ordering isn't supported by the backend.")
+
+            ascending = not order.startswith("-")
+            if not self.query.standard_ordering:
+                ascending = not ascending
+
+            name = order.lstrip("+-")
+            if name == "pk":
+                name = opts.pk.name
+
+            field_ordering.append((opts.get_field(name), ascending))
+        return field_ordering
+
+    def get_collection(self):
+        return self.connection.get_collection(self.query.get_meta().db_table)
+
+
+class SQLInsertCompiler(SQLCompiler):
+    def execute_sql(self, returning_fields=None):
+        self.pre_sql_setup()
+
+        to_insert = []
+        # pk_field = self.query.get_meta().pk
+        keys = []
+        for obj in self.query.objs:
+            field_values = {}
+            for field in self.query.fields:
+                value = field.get_db_prep_save(
+                    getattr(obj, field.attname)
+                    if self.query.raw
+                    else field.pre_save(obj, obj._state.adding),
+                    connection=self.connection,
+                )
+                if value is None and not field.null and not field.primary_key:
+                    raise IntegrityError(
+                        "You can't set %s (a non-nullable field) to None." % field.name
+                    )
+
+                field_values[field.column] = value
+            to_insert.append(field_values)
+
+            # TODO: pass the key value through db converters (use pk_field).
+            keys.append(self.insert(to_insert, returning_fields=returning_fields))
+
+        return keys
+
+    @safe_call
+    def insert(self, docs, returning_fields=None):
+        """
+        Store a document using field columns as element names.
+
+        If just a {pk_field: None} mapping is given a new empty document is
+        created, otherwise value for a primary key may not be None.
+        """
+        for doc in docs:
+            if doc.get("_id", NOT_PROVIDED) is None:
+                if len(doc) == 1:
+                    # insert with empty model
+                    doc.clear()
+                else:
+                    raise DatabaseError("Can't save entity with _id set to None")
```

Comment:
Comment one: This code seems to have been obsoleted since [its introduction](https://github.com/django-nonrel/mongodb-engine/commit/7843ad1cb5d6ed4aa3c7460a83ebbb6efe281c1f). The test now passes without it.

Comment two: Django doesn't support [composite primary keys](https://code.djangoproject.com/ticket/373).